File: src/Wording.Core/Packs/WordPackReader.cs
using System.Text.Json;
using Wording.Core.Storage;

namespace Wording.Core.Packs;

/// <summary>
/// Parses a downloaded pack and decides whether it is fit to import.
/// <para>
/// Every check happens here, before anything reaches the disk: the caller gets either a
/// pack that is known to be safe to write, or a <see cref="WordPackException"/>.
/// </para>
/// <para>
/// Structural problems are refused. Two display-only fields - the name and the
/// description - are truncated instead, because they cannot harm anything and failing a
/// whole pack over a long title would leave the user with no way forward: they did not
/// write the file and cannot fix it.
/// </para>
/// </summary>
public static class WordPackReader
{
    public static WordPack Read(ReadOnlySpan<byte> payload)
    {
        if (payload.Length > PackLimits.MaxPayloadBytes)
        {
            throw new WordPackException(
                PackProblem.TooLarge,
                $"the pack is {payload.Length} bytes, the limit is {PackLimits.MaxPayloadBytes}");
        }

        WordPack? pack;

        try
        {
            pack = JsonSerializer.Deserialize(payload, WordJsonContext.Default.WordPack);
        }
        catch (JsonException error)
        {
            throw new WordPackException(PackProblem.Malformed, "the pack is not valid JSON", error);
        }

        if (pack is null)
        {
            throw new WordPackException(PackProblem.Malformed, "the pack is empty");
        }

        return Validate(pack);
    }

    /// <summary>
    /// Applies every rule to an already-parsed pack. Split out so the repository's own
    /// packs can be checked without going through a download.
    /// </summary>
    public static WordPack Validate(WordPack pack)
    {
        ArgumentNullException.ThrowIfNull(pack);

        var slug = PackSlug.Require(pack.Id);
        var name = Clean(pack.Name);

        if (name.Length == 0)
        {
            throw new WordPackException(PackProblem.Malformed, "the pack has no name");
        }

        if (pack.Words.Count > PackLimits.MaxWords)
        {
            throw new WordPackException(
                PackProblem.TooLarge,
                $"the pack has {pack.Words.Count} words, the limit is {PackLimits.MaxWords}");
        }

        var entries = new List<PackEntry>(pack.Words.Count);

        foreach (var entry in pack.Words)
        {
            var original = Clean(entry.Original);
            var translation = Clean(entry.Translation);

            // A blank line in a hand-edited pack is noise, not a reason to refuse the
            // rest of it.
            if (original.Length == 0 || translation.Length == 0)
            {
                continue;
            }

            // A field this long is not a word - it is a sign the file is some other
            // format that happens to parse. Truncating would silently change meaning.
            if (original.Length > PackLimits.MaxFieldLength || translation.Length > PackLimits.MaxFieldLength)
            {
                throw new WordPackException(
                    PackProblem.Malformed,
                    $"'{Preview(original)}' exceeds the {PackLimits.MaxFieldLength} character limit for a word");
            }

            entries.Add(new PackEntry { Original = original, Translation = translation });
        }

        if (entries.Count == 0)
        {
            throw new WordPackException(PackProblem.Empty, "the pack carries no usable word");
        }

        return new WordPack
        {
            Id = slug,
            Name = Truncate(name, PackLimits.MaxNameLength),
            Kind = PackKind.Normalize(pack.Kind),
            Description = Truncate(Clean(pack.Description ?? string.Empty), PackLimits.MaxDescriptionLength) is { Length: > 0 } text
                ? text
                : null,
            Words = entries,
        };
    }

    /// <summary>
    /// Trims, and folds every control character - newlines and tabs included - into a
    /// space. They would otherwise reach a notification body and a grid cell.
    /// </summary>
    static string Clean(string value)
    {
        if (string.IsNullOrEmpty(value))
        {
            return string.Empty;
        }

        var characters = new char[value.Length];

        for (var i = 0; i < value.Length; i++)
        {
            characters[i] = char.IsControl(value[i]) ? ' ' : value[i];
        }

        return new string(characters).Trim();
    }

    static string Truncate(string value, int limit) =>
        value.Length <= limit ? value : value[..limit].TrimEnd();

    static string Preview(string value) => Truncate(value, 40);
}

File: macos/Sources/WordingKit/WordPackReader.swift
import Foundation

/// Parses a downloaded pack and decides whether it is fit to import.
///
/// Every check happens here, before anything reaches the disk: the caller gets either a
/// pack that is known to be safe to write, or a `WordPackError`.
///
/// Structural problems are refused. Two display-only fields - the name and the
/// description - are truncated instead, because they cannot harm anything and failing a
/// whole pack over a long title would leave the user with no way forward: they did not
/// write the file and cannot fix it.
///
/// A port of `Wording.Core.Packs.WordPackReader`.
public enum WordPackReader {
    public static func read(_ payload: Data) throws -> WordPack {
        guard payload.count <= PackLimits.maxPayloadBytes else { throw WordPackError.tooLarge }

        let pack: WordPack

        do {
            pack = try JSONDecoder().decode(WordPack.self, from: payload)
        } catch {
            throw WordPackError.malformed("the pack is not valid JSON")
        }

        return try validate(pack)
    }

    /// Applies every rule to an already-parsed pack. Split out so the repository's own
    /// packs can be checked without going through a download.
    public static func validate(_ pack: WordPack) throws -> WordPack {
        let slug = try PackSlug.require(pack.id)
        let name = clean(pack.name)

        guard !name.isEmpty else { throw WordPackError.malformed("the pack has no name") }
        guard pack.words.count <= PackLimits.maxWords else { throw WordPackError.tooLarge }

        var entries: [PackEntry] = []
        entries.reserveCapacity(pack.words.count)

        for entry in pack.words {
            let original = clean(entry.original)
            let translation = clean(entry.translation)

            // A blank line in a hand-edited pack is noise, not a reason to refuse the
            // rest of it.
            if original.isEmpty || translation.isEmpty { continue }

            // A field this long is not a word - it is a sign the file is some other
            // format that happens to parse. Truncating would silently change meaning.
            guard original.count <= PackLimits.maxFieldLength,
                translation.count <= PackLimits.maxFieldLength
            else {
                throw WordPackError.malformed(
                    "a word exceeds the \(PackLimits.maxFieldLength) character limit")
            }

            entries.append(PackEntry(original: original, translation: translation))
        }

        guard !entries.isEmpty else { throw WordPackError.empty }

        let description = truncate(clean(pack.description ?? ""), to: PackLimits.maxDescriptionLength)

        return WordPack(
            id: slug,
            name: truncate(name, to: PackLimits.maxNameLength),
            description: description.isEmpty ? nil : description,
            words: entries
        )
    }

    /// Trims, and folds every control character - newlines and tabs included - into a
    /// space. They would otherwise reach a notification body and a list row.
    private static func clean(_ value: String) -> String {
        String(value.map { $0.isNewline || $0.unicodeScalars.allSatisfy(CharacterSet.controlCharacters.contains) ? " " : $0 })
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func truncate(_ value: String, to limit: Int) -> String {
        guard value.count > limit else { return value }

        return String(value.prefix(limit)).trimmingCharacters(in: .whitespaces)
    }
}
